packages/jest-runner/worker.ts

```diff
@@ -215,15 +232,91 @@
 		});
 	}
 
-	// noinspection JSUnusedLocalSymbols
 	private async runHooks(
 		hookType: string,
-		block: Circus.DescribeBlock | Circus.TestEntry,
+		block: Circus.DescribeBlock | Circus.TestEntry, // TODO: kill testEntry
 		ancestors: string[],
-		// eslint-disable-next-line @typescript-eslint/no-unused-vars
 		shouldRunInAncestors = false
 	) {
-		// TODO: Implement
+		let correctHooks: Circus.Hook[] = [];
+		if (block.type === "describeBlock") {
+			if (shouldRunInAncestors && block.parent) {
+				correctHooks = block.parent.hooks;
+			} else {
+				correctHooks = block.hooks;
+			}
+		}
+		for (const hook of correctHooks) {
+			if (hook.type === hookType) {
+				const timeout = hook.timeout ?? this.defaultTimeout;
+				const hookFn = new Promise((resolve, reject) => {
+					setTimeout(() => {
+						reject(
+							this.adjustErrorStack(
+								new Error(
+									`Exceeded timeout of ${timeout} ms for a hook.\nIncrease the timeout value, if this is a long-running test.`
+								),
+								hook.asyncError.stack
+							)
+						);
+					}, timeout);
+
+					let doneMessage: string | Error | undefined;
+					let wasDoneCalled = false;
+					let doneMessageStack: string | undefined;
+					const result = hook.fn((doneMsg?: string | Error) => {
+						doneMessage = doneMsg;
+						wasDoneCalled = true;
+						doneMessageStack = new Error().stack;
+					});
+
+					if (result instanceof Promise) {
+						if (hook.fn.length === 0) {
+							result.then(resolve, reject);
+						} else {
+							reject(
+								this.adjustErrorStack(
+									new Error(
+										"Using done callback in async test is not allowed."
+									),
+									hook.asyncError.stack
+								)
+							);
+						}
+					} else if (isGeneratorFunction(result)) {
+						reject(
+							this.adjustErrorStack(
+								new Error(
+									`Generators are currently not supported by Jazzer.js`
+								),
+								hook.asyncError.stack
+							)
+						);
+					} else if (hook.fn.length === 0) {
+						// hook function has no arguments
+						resolve(result);
+					} else {
+						// hook function expects a done callback
+						// before* or after* functions that use a done callback
+						// will either: 1) timeout; 2) call the done callback
+						if (wasDoneCalled) {
```

Comment:
I think a test case calling `done` in an async fashion is missing. `wasDoneCalled` is only checked in the synchronous case, or am I missing something?

Reply:
The check is done when we handle the promise result, a bit higher.
```
if (result instanceof Promise) {
      if (hook.fn.length === 0) {
          result.then(resolve, reject);
      } else {
```

Reply:
In case a `done` callback is used there is only an `undefined` result and hence no waiting for the callback invocation.

Reply:
Added another promise-layer to wait for done callback invocation.